optax/contrib/_sophia_h.py

```diff
@@ -0,0 +1,307 @@
+# Copyright 2023 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Sophia-H optimizer.
+
+A contributed implementation of the Sophia-H optimizer from "Sophia: A Scalable
+Stochastic Second-order Optimizer for Language Model Pre-training"
+(https://arxiv.org/abs/2305.14342) by Hong Liu, Zhiyuan Li, David Hall,
+Percy Liang, and Tengyu Ma.
+
+This contribution is heavily based on the implementation of Sophia by levanter
+(https://github.com/stanford-crfm/levanter) with some changes.
+"""
+from typing import Any, NamedTuple, Optional, Union, Callable
+
+import jax
+from jax import numpy as jnp
+from jax.random import PRNGKey
+
+from optax._src import base
+from optax._src import transform
+from optax._src.numerics import safe_int32_increment
+from optax._src.utils import canonicalize_dtype
+from optax._src.combine import chain
+from optax import tree_utils as otu
+
+
+class SophiaHState(NamedTuple):
+  """State for Sophia-H and similar."""
```

Comment:
What does similar refer to?

Reply:
Just a line from levanter, I can change it